ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/OngoingAuthorizationObserver.scala

```diff
@@ -29,12 +63,85 @@ private[auth] final class OngoingAuthorizationObserver[A](
   override def setMessageCompression(b: Boolean): Unit = observer.setMessageCompression(b)
 
   override def onNext(v: A): Unit =
-    authorized(claims) match {
+    authorize match {
+      case _ if aborted.get() => ()
       case Right(_) => observer.onNext(v)
-      case Left(authorizationError) => observer.onError(throwOnFailure(authorizationError))
+      case Left(authorizationError: AuthorizationError) =>
+        val e: StatusRuntimeException =
+          errorFactories.permissionDenied(authorizationError.reason)(errorLogger)
+        observer.onError(e)
     }
 
   override def onError(throwable: Throwable): Unit = observer.onError(throwable)
 
   override def onCompleted(): Unit = observer.onCompleted()
+
+  private def authorize: Either[AuthorizationError, Unit] = {
+    val now = nowF()
+    for {
+      _ <- originalClaims.notExpired(now)
+      _ = validateClaimsResolvedFromUser(now)
+    } yield {
+      ()
+    }
+  }
+
+  /** Aborts the stream by throwing an exception
+    * if any change in compared the original user claims has been detected
+    * or if user claims change check times out.
+    */
+  private def validateClaimsResolvedFromUser(now: Instant): Unit = {
+    if (originalClaims.resolvedFromUser) {
+      if (shouldAbort.get()) {
+        signalError()
+      } else if (
+        !userRightsRefreshInProgress.get()
+        && now.isAfter(
+          lastUserInfoAccessTime.plus(claimsFreshnessCheckDelayInSeconds.toLong, ChronoUnit.SECONDS)
+        )
+      ) {
+        userRightsRefreshInProgress.set(true)
+        scheduleAuthenticationRefresh()
+      } else if (userRightsRefreshInProgress.get()) {
+        // Timing out user rights refresh
+        if (
+          now.isAfter(
+            lastUserInfoAccessTime
+              .plus(2 * claimsFreshnessCheckDelayInSeconds.toLong, ChronoUnit.SECONDS)
+          )
+        ) {
+          shouldAbort.set(true)
+          signalError()
+        }
+      }
+
+    }
+  }
+
+  private def scheduleAuthenticationRefresh(): Unit = {
+    userManagementStore
+      .listUserRights(userId)
+      .onComplete {
+        case Failure(_) => shouldAbort.set(true)
+        case Success(Left(_)) => shouldAbort.set(true)
+        case Success(Right(userRights)) =>
+          if (!shouldAbort.get()) {
+            val updatedClaims = AuthorizationInterceptor.convertUserRightsToClaims(userRights)
+            shouldAbort.set(updatedClaims.toSet != originalClaims.claims.toSet)
```

Comment:
I tried to guarantee that by checking `if (!shouldAbort.get()) {` above.
It should work assuming at most one rights' refresh is scheduled at a time which should hold (through my use of atomic types) as long as `StreamObserver.onNext(v)` is called serially- which I think it is as I found this pieces of documentation:
1)
`io.grpc.stub.StreamObserver`:
> Since individual StreamObservers are not thread-safe, if multiple threads will be writing to a StreamObserver concurrently, the application must synchronize calls.
2)
`com.daml.grpc.adapter.server.akka.ServerAdapter` bridges from Akka's source to grpc stream observer via  `org.reactivestreams.Subscriber` and reactive spec says:
> onSubscribe, onNext, onError and onComplete signaled to a Subscriber MUST be signaled serially.
https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.3/README.md#1.3

where `serially` is defined as
> In the context of a Signal, non-overlapping. In the context of the JVM, calls to methods on an object are serial if and only if there is a happens-before relationship between those calls (implying also that the calls do not overlap). When the calls are performed asynchronously, coordination to establish the happens-before relationship is to be implemented using techniques such as, but not limited to, atomics, monitors, or locks.

Reply:
Changed approach to use Akka scheduler and calling `observer.onError` (in a `synchronized` block) from the scheduled tasks.